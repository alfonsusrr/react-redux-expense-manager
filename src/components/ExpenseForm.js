import React from "react";
import moment from 'moment'
import DatePicker from 'react-date-picker'

const now = moment()
class ExpenseForm extends React.Component {
    state = this.props.expense ? {
        description: this.props.expense.description, 
        note: this.props.expense.note, 
        amount: this.props.expense.amount, 
        createdAt: moment(new Date(this.props.expense.createdAt)), 
        error: ''
    } : {
        description: '',
        note: '',
        amount: '',
        createdAt: moment(),
        error: ''
    }
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value

        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (date, e) => {
        if (date) {
            this.setState(() => ({ createdAt: moment(date)}))
        } else {
            this.setState(() => ({ createdAt: moment(new Date())}))
        }
    }
    onSubmit = (e) => {
        e.preventDefault()
        
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {!!this.state.error ? <p>{this.state.error}</p> : undefined}
                    <input
                        data-testid="description"
                        type="text"
                        placeholder="Description"
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    ></input>
                    <input
                        type="text"
                        data-testid="amount"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    ></input>
                    <DatePicker
                        data-testid="created-at"
                        value={this.state.createdAt.toDate()}
                        onChange={this.onDateChange}
                    />
                    <textarea
                        data-testid="note"
                        placeholder="Add a note for your expense (optional)"
                        onChange={this.onNoteChange}
                        value={this.state.note}
                    ></textarea>
                    <button data-testid="submit" >{this.props.action} Expense</button>
                </form>
            </div>
        )
    }
}

export default ExpenseForm