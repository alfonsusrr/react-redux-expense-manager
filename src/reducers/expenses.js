import { createSlice } from '@reduxjs/toolkit';
import { v4 as uuid } from 'uuid';

const expensesSlice = createSlice({
    name: 'expenses',
    initialState: [],
    reducers: {
        addExpense: {
            reducer: (state, action) => {
                state.push(action.payload)
                return state
            },
            prepare: ({ description = '', note = '', amount = 0, createdAt = (new Date()).getTime()}) => ({
                payload: {
                    id: uuid(),
                    description,
                    note,
                    amount,
                    createdAt
                }
            })
        },
        removeExpense: {
            reducer: (state, action) => {
                return state.filter((expense) => expense.id !== action.payload.id)
            },
            prepare: ({ id }) => ({
                payload: {
                    id
                }
            })
        },
        editExpense: {
            reducer: (state, action) => {
                return state.map((expense) => (expense.id === action.payload.id) ? { ...expense, ...action.payload.updates  } : expense)
            },
            prepare: ({ id, updates }) => ({
                payload: { 
                    id, 
                    updates 
                }
            })
        }   
    }
})

export const { addExpense, removeExpense, editExpense } = expensesSlice.actions
export default expensesSlice.reducer