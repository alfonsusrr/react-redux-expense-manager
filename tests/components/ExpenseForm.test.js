/**
 * @jest-environment jsdom
 */

import React from 'react';
import '@testing-library/jest-dom'
import { screen, fireEvent, getByTestId, render } from '@testing-library/react'
import renderWithProviders from './utils/render';
import ExpenseForm from '../../src/components/ExpenseForm'
import { emptyStore } from '../fixtures/testStore';
import expenses from '../fixtures/expenses';
import moment from 'moment'

describe('Expense form', () => {
    test('should render ExpenseForm correctly', () => {
        const { asFragment } = renderWithProviders(<ExpenseForm/>, { store: emptyStore })
        expect(asFragment()).toMatchSnapshot()
    })

    test('should correctly render add expense form', () => {
        renderWithProviders(<ExpenseForm action="Add"/>, { store: emptyStore })
        expect(screen.getByText("Add Expense")).toBeInTheDocument()
    })
    
    test('should correctly render edit expense form with expense data', () => {
        renderWithProviders(<ExpenseForm expense={expenses[0]} action="Edit"/>, { store: emptyStore})
        expect(screen.getByTestId("description")).toHaveValue(expenses[0].description)
        expect(screen.getByTestId("amount")).toHaveValue(expenses[0].amount.toString())
        expect(screen.getByTestId("note")).toHaveValue(expenses[0].note)
        expect(screen.getByDisplayValue(moment(expenses[0].createdAt).format("YYYY-MM-DD"))).toBeInTheDocument()
    })

    test('should render error for invalid form submission', () => {
        renderWithProviders(<ExpenseForm />, { store: emptyStore })
        fireEvent.change(screen.getByTestId("description", { target: {value: 'Test description'}}))
        fireEvent.click(screen.getByTestId("submit"))
        expect(screen.getByText('Please provide description and amount')).toBeInTheDocument()
        fireEvent.change(screen.getByTestId("description", { target: {value: ''}}))
        fireEvent.change(screen.getByTestId("amount", { target: {value: '2020'}}))
        fireEvent.click(screen.getByTestId("submit"))
        expect(screen.getByText('Please provide description and amount')).toBeInTheDocument()
    })

    test('should not set amount other than number', () => {
        const amount = '1as21'
        renderWithProviders(<ExpenseForm />, { store: emptyStore })
        fireEvent.change(screen.getByTestId("amount", { target: {value: amount}}))
        expect(screen.getByTestId("amount")).toHaveValue('')
    })

    test('should set amount if valid number', () => {
        renderWithProviders(<ExpenseForm />, { store: emptyStore })
        fireEvent.change(screen.getByTestId("amount", { target: { value: '2020' }}))
        expect(screen.getByTestId("amount")).toHaveValue('2020')
    })

    test('should set description', () => {
        renderWithProviders(<ExpenseForm />, { store: emptyStore })
        fireEvent.change(screen.getByTestId("description", { target: { value: 'test' }}))
        expect(screen.getByTestId("description")).toHaveValue('test')
    })
});