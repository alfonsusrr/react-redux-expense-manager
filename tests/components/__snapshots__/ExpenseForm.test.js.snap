// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense form should render ExpenseForm correctly 1`] = `
<DocumentFragment>
  <div>
    <form>
      <input
        data-testid="description"
        placeholder="Description"
        type="text"
        value=""
      />
      <input
        data-testid="amount"
        placeholder="Amount"
        type="text"
        value=""
      />
      <div
        class="react-date-picker react-date-picker--closed react-date-picker--enabled"
      >
        <div
          class="react-date-picker__wrapper"
        >
          <div
            class="react-date-picker__inputGroup"
          >
            <input
              max="275760-09-13"
              min="0001-01-01"
              name="date"
              style="visibility: hidden; position: absolute; z-index: -999;"
              type="date"
              value="2022-07-05"
            />
            <input
              autocomplete="off"
              class="react-date-picker__inputGroup__input react-date-picker__inputGroup__month"
              data-input="true"
              inputmode="numeric"
              max="12"
              min="1"
              name="month"
              placeholder="--"
              style="width: 1px;"
              type="number"
              value="7"
            />
            <span
              class="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autocomplete="off"
              class="react-date-picker__inputGroup__input react-date-picker__inputGroup__day"
              data-input="true"
              inputmode="numeric"
              max="31"
              min="1"
              name="day"
              placeholder="--"
              style="width: 1px;"
              type="number"
              value="5"
            />
            <span
              class="react-date-picker__inputGroup__divider"
            >
              /
            </span>
            <input
              autocomplete="off"
              class="react-date-picker__inputGroup__input react-date-picker__inputGroup__year"
              data-input="true"
              inputmode="numeric"
              max="275760"
              min="1"
              name="year"
              placeholder="----"
              step="1"
              style="width: 4px;"
              type="number"
              value="2022"
            />
          </div>
          <button
            class="react-date-picker__clear-button react-date-picker__button"
            type="button"
          >
            <svg
              class="react-date-picker__clear-button__icon react-date-picker__button__icon"
              height="19"
              stroke="black"
              stroke-width="2"
              viewBox="0 0 19 19"
              width="19"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="4"
                x2="15"
                y1="4"
                y2="15"
              />
              <line
                x1="15"
                x2="4"
                y1="4"
                y2="15"
              />
            </svg>
          </button>
          <button
            class="react-date-picker__calendar-button react-date-picker__button"
            type="button"
          >
            <svg
              class="react-date-picker__calendar-button__icon react-date-picker__button__icon"
              height="19"
              stroke="black"
              stroke-width="2"
              viewBox="0 0 19 19"
              width="19"
              xmlns="http://www.w3.org/2000/svg"
            >
              <rect
                fill="none"
                height="15"
                width="15"
                x="2"
                y="2"
              />
              <line
                x1="6"
                x2="6"
                y1="0"
                y2="4"
              />
              <line
                x1="13"
                x2="13"
                y1="0"
                y2="4"
              />
            </svg>
          </button>
        </div>
      </div>
      <textarea
        data-testid="note"
        placeholder="Add a note for your expense (optional)"
      />
      <button
        data-testid="submit"
      >
         Expense
      </button>
    </form>
  </div>
</DocumentFragment>
`;
